### ISSUE TITLE
Faster concurrently

### ISSUE THREAD

**Issue by basvandijk** [simonmar/async#5]
I was wondering if the following implementation of `concurrently` has been considered before:

``` haskell
concurrently :: IO a -> IO b -> IO (a,b)
concurrently left right = do
  mv <- newEmptyMVar
  myTid <- myThreadId
  mask $ \restore -> do
    tid <- forkIO $
             (restore right `catchAll`
                (\e -> throwTo myTid e >> return undefined)) >>= putMVar mv
    l <- restore left `onException` killThread tid
    r <- takeMVar mv
    return (l, r)
```

This implementation forks one thread instead of two.

It's 50% faster than the current `concurrently` in the `concasync` benchmark.

I expect to also see a significant efficiency boost of the `traverse` method of  `Concurrently` which is used by `mapConcurrently`. Currently `mapConcurrently` forks 2n threads for a list of lengh n. Using this `concurrently` it only forks n threads.

The only thing I'm not yet sure about (and the reason I haven't created a pull request yet) is what happens when both `left` and `right` throw an exception.

**Comment by snoyberg** [comment]
Recommendation: instead of catching just an `AsyncException`, check if the type casts to `SomeAsyncException`, [e.g.](https://github.com/fpco/safe-exceptions/blob/2ec1dc69b30e29782934b7bb7d7f4f383812cb8f/src/Control/Exception/Safe.hs#L329):

```haskell
isAsyncException :: Exception e => e -> Bool
isAsyncException e =
    case fromException (toException e) of
        Just (SomeAsyncException _) -> True
        Nothing -> False
```

This requires base 4.7 or later. If support for earlier base is desired, I'd say use CPP to provide a fallback to just using `AsyncException`, or consider using a much more exhaustive test such as:

https://www.stackage.org/haddock/lts-6.4/unexceptionalio-0.3.0/src/UnexceptionalIO.html#syncIO

Though looking at that source, I'm not convinced that list is correct.

**Comment by simonmar** [comment]
Yes, I'd pretty much decided to do that, just hadn't got around to it.

**Comment by basvandijk** [comment]
Yes, I also [do that](https://github.com/simonmar/async/pull/21/files#diff-54890a66957e2efe3f5ceaa39499ff45R237) in #21.

**Comment by Saizan** [comment]
Regarding the impact on `mapConcurrently` you would actually like to `forkIO` the `left` action, as the `right` is presumably the one recursing into the list.
This would make it so the whole list of threads could get forked within the currently running one, instead of waiting for each child to get scheduled in turn.

Also it would allow each element thread to independently `putMVar` as soon as they are done, as opposed to waiting for the `takeMVar` with the tail to unblock so we can `putMVar` the parent cons cell, which seems it would cause less friction?

Of course `traverse` itself could switch the order of computations, but that'd be pretty unnatural.

Given how old the issue is I guess it's not been so pressing, I got bit by this in a different library modeled after `async`, where having nested children threads is much more costly than a flat structure.
